lib/entities.ts

```diff
@@ -1,10 +1,126 @@
 import { getCollection } from "./collection.js";
-import { HassEntities, StateChangedEvent, UnsubscribeFunc } from "./types.js";
+import {
+  Context,
+  HassEntities,
+  StateChangedEvent,
+  UnsubscribeFunc,
+} from "./types.js";
 import { Connection } from "./connection.js";
 import { Store } from "./store.js";
 import { getStates } from "./commands.js";
+import { atLeastHaVersion } from "./util.js";
 
-function processEvent(store: Store<HassEntities>, event: StateChangedEvent) {
+interface EntityState {
+  /** state */
+  s: string;
+  /** attributes */
+  a: { [key: string]: any };
+  /** context */
+  c: Context | string;
+  /** last_changed; if set, also applies to lu */
+  lc: number;
+  /** last_updated */
+  lu: number;
+}
+
+interface EntityDiff {
+  /** additions */
+  "+"?: Partial<EntityState>;
+  /** subtractions */
+  "-"?: Pick<EntityState, "a">;
+}
+
+interface StatesUpdates {
+  /** add */
+  a?: Record<string, EntityState>;
+  /** remove */
+  r?: string[]; // remove
+  /** change */
+  c: Record<string, EntityDiff>;
+}
+
+function processEvent(store: Store<HassEntities>, updates: StatesUpdates) {
+  const state = { ...store.state };
+
+  if (updates.a) {
+    for (const entityId in updates.a) {
+      const newState = updates.a[entityId];
+      let last_changed = new Date(newState.lc * 1000).toISOString();
+      state[entityId] = {
+        entity_id: entityId,
+        state: newState.s,
+        attributes: newState.a,
+        context: (typeof newState.c === 'string') ? {"id": newState.c, "parent_id": null, "user_id": null} : newState.c,
+        last_changed: last_changed,
+        last_updated: newState.lu ? new Date(newState.lu * 1000).toISOString() : last_changed,
+      };
+    }
+  }
+
+  if (updates.r) {
+    for (const entityId of updates.r) {
+      delete state[entityId];
+    }
+  }
+
+  if (updates.c) {
+    for (const entityId in updates.c) {
+      let entityState = state[entityId];
+
+      if (!entityState) {
+        console.warn("Received state update for unknown entity", entityId);
+        continue;
+      }
+
+      entityState = { ...entityState };
+
+      const { "+": toAdd, "-": toRemove } = updates.c[entityId];
+
+      if (toAdd) {
+        if (toAdd.s) {
+          entityState.state = toAdd.s;
+        }
+        if (toAdd.c) {
+          if (typeof toAdd.c === 'string') {
+            entityState.context = { ...entityState.context, "id": toAdd.c }
+          } else {
+            entityState.context = { ...entityState.context, ...toAdd.c }
+          }
+        }
+        if (toAdd.lc) {
+          entityState.last_updated = entityState.last_changed = new Date(toAdd.lc * 1000).toISOString();
```

Comment:
The irony is that every time this value is actually used, we have to convert it back to a `Date`. In a future PR we might expand `HassEntity` to contain `_dt` versions

Reply:
That a good idea. round tripping toISOString on that many can't be cheap, but at least its way faster than python AFAICT